server/controllers/upload-controller.js

```diff
@@ -0,0 +1,312 @@
+const knex = require("./../db");
+const fs = require("fs");
+const csv = require("fast-csv");
+const moment = require("moment");
+const bcrypt = require("bcrypt");
+
+const clearFile = async (path) => {
+  return new Promise(async (resolve, reject) => {
+    try {
+      fs.unlinkSync(path);
+    } catch {
+      if (fs.existsSync(path)) {
+        clearFile(path);
+      }
+    }
+    if (!fs.existsSync(path)) resolve();
+    else reject();
+  });
+};
+
+const checkSupportOfficer = async (userId) => {
+  const user = await knex
+    .select({ role: "role" })
+    .from("user")
+    .where("id", userId)
+    .catch(() => {
+      throw { msg: `There was an error retrieving user info`, status: 400 };
+    });
+  if (user.length == 1 && !["supportOfficer"].includes(user[0].role)) {
+    return false;
+  }
+  return true;
+};
+/*
+const checkDuplicate = async (values) => {  //check duplicate elements (same id)
+  //console.log(values)
+  let valueArr = values.map(function(item){ return item.id });
+  valueArr.some(function(item, idx){ 
+    console.log(valueArr.indexOf(item) != idx)
+    return valueArr.indexOf(item) != idx    //return true if ther is duplicate
+});
+}*/
+
+const readFile = async (userId, path, type, semesterId) => {
+  //type: student, teacher, course, enrollment, schedule
+  const headers = {
+    student: ["id", "name", "surname", "city", "email", "birthday", "ssn"],
+    teacher: ["id", "name", "surname", "email", "ssn"],
+    course: ["id", "year", "semester", "name", "main_prof"],
+    enrollment: ["course_id", "student_id"],
+    schedule: ["id", "room", "day", "capacity", "time"],
+  };
+  const tables = {
+    student: "user",
+    teacher: "user",
+    course: "course",
+    enrollment: "course_available_student",
+    schedule: "lecture",
+  };
+  return new Promise(async (resolve, reject) => {
+    if (!(await checkSupportOfficer(userId))) {
+      reject({
+        msg: "Only support officer can upload files.",
+        status: 401,
+      });
+      return;
+    }
+    let rows = [];
+    var tot = 0;
+    const stream = fs
+      .createReadStream(path)
+      .pipe(
+        csv.parse({
+          headers: headers[type],
+          renameHeaders: true,
+        })
+      )
+      .on("error", (error) => {
+        reject({
+          msg: error,
+        });
+      })
+      .on("data", async (row) => {
+        stream.pause();
+        //console.log("STEP 0n")
+        if (["student", "teacher"].includes(type)) {
+          row.password_hash = bcrypt.hashSync("password", 1);
+          row.role = type;
+        }
+        if (type === "schedule") {
+          try {
+            const count = await generateLectures(semesterId, row);
+            //console.log("STEP 2n DENTRO READ FILE",count);
```

Comment:
Usually it's seen as a bad practice to commit files containing commented lines of code.
Also, please don't use Italian words.